### DISCUSSION TITLE
Using Solidjs with P5js (in Astro)

### DISCUSSION THREAD

**Discussion by ff6347** [solidjs/solid#1704] [Q&A] [answered]
Hi, 

for a small project I wanted to give Solidjs with Astro a try but I'm failing with a basic logic when something is rendered. 

I have a component called Sketch

```tsx
import { createSignal, createEffect, onCleanup } from "solid-js";
import p5 from "p5";

const Sketch = () => {
	let canvasRef: any;
	let x = 50;
	let y = 50;

	const [width, setWidth] = createSignal(200);
	const [height, setHeight] = createSignal(200);



	// create and manage p5js instance
	createEffect(() => {
		const sketch = (p: p5) => {
			p.setup = () => {
				p.createCanvas(width(), height());
			};
			p.draw = () => {
				p.background(0);
				p.fill(0);
				p.ellipse(x, y, 50, 50);
			};
		};

		const p5canvas = new p5(sketch, canvasRef);
		// clean up when component is unmounted
		onCleanup(() => {
			p5canvas.remove();
		});
	}, [canvasRef]);

	return <canvas ref={canvasRef} />;
};

export default Sketch;

```

And my index.astro where the component should be rendered.

```
---
import Sketch from "../components/Sketch";
// Component Imports

// Full Astro Component Syntax:
// https://docs.astro.build/core-concepts/astro-components/
---

<html lang="en">
	<head>
		<meta charset="utf-8" />
		<meta name="viewport" content="width=device-width" />
		<meta name="generator" content={Astro.generator} />
		<link rel="icon" type="image/svg+xml" href="/favicon.svg" />
		<style>
			html,
			body {
				font-family: system-ui;
				margin: 0;
			}
			body {
				padding: 2rem;
			}
		</style>
	</head>
	<body>
		<main>
			<div id="app" >
				<Sketch client:only="solid-js"/>
			</div>

		</main>
	</body>
</html>

```

This is all pretty basic. The page renders and also the markup for the astro island is created. The inspector shows me the canvas with the full black background when I hover over hit, but the page itself does not show it. 

My guess is that the p5js library is creating the canvas after Solid/Astro renders the canvas.

Any guesses or hints would be highly appreciated

**Comment by snnsnn** [comment, marked as answer]
You can use functional ref for rendering p5js rather than an effect. See this answer for how it is done: https://github.com/solidjs/solid/discussions/1694

**Reply by ff6347** [reply]
Thank you. It works now. For future reference. I had to not create a canvas but a div element since p5js creates the canvas by itself and I can make it a child of a div. Here is the working code

```ts
import { createSignal } from "solid-js";
import p5 from "p5";

const Sketch = () => {
	let x = 50;
	let y = 50;

	const [width, setWidth] = createSignal(200);
	const [height, setHeight] = createSignal(200);

	const createSketch = (ref: HTMLDivElement) => {
		const sketch = (p: p5) => {
			p.setup = () => {
				const canvas = p.createCanvas(width(), height());
				canvas.parent(ref);
			};
			p.draw = () => {
				p.background(0);
				p.fill(0);
				p.ellipse(x, y, 50, 50);
			};
		};
		new p5(sketch, ref);
	};

	return <div ref={(el) => createSketch(el)} />;
};

export default Sketch;

```

**Reply by snnsnn** [reply]
The ref callback is redundant:

```tsx
<div ref={(el) => createSketch(el)} />
```

You can achieve the same result by passing `createSketch`:

```tsx
<div ref={createSketch} />
```

**Reply by AlexErrant** [reply]
^ This is called an [eta reduction](http://wiki.haskell.org/Eta_conversion) in functional programming circles. Note that [it only works for one-argument functions](https://stackoverflow.com/q/12344087) because Javascript doesn't do automatic currying.

**Reply by snnsnn** [reply]
> ^ This is called an [eta reduction](http://wiki.haskell.org/Eta_conversion) in functional programming circles.

Sorry I could not get your point, compiled code calls the provided callback internally, so it has nothing to do with currying.

**Reply by AlexErrant** [reply]
Sorry, I'm not contradicting you, I'm just adding extra info. The act of transforming `(el) => createSketch(el)` to `createSketch` is called eta reduction. However, in Javascript, it only works with functions with one argument, which is the case in this question, so no big deal. However, _if_ it had more than one argument, then you'd run into issues like [the linked Stackoverflow](https://stackoverflow.com/q/12344087), and this is due to limitations in Javascript-the-language. Other languages like Haskell and F#, which support function currying by default don't have this problem with multi-arity functions (because due to currying-by-default, _every_ function is single-arity). To be clear, this isn't me dunking on Javascript; currying-by-default has tradeoffs.

This is a big aside and I regret raising it 😅
